c++/src/kj/mutex.c++

```diff
@@ -58,7 +59,100 @@
 #include <windows.h>
 #endif
 
+#if KJ_TRACK_LOCK_BLOCKING
+#include <pthread.h>
+#endif
+
 namespace kj {
+
+#if KJ_TRACK_LOCK_BLOCKING
+static pthread_key_t waitingTypeKey;
```

Comment:
I agree I too would strongly prefer thread_local. I've read too much conflicting information that I don't think there's a clear answer & basically amounts to "It depends. YOLO". If TLS space is lazily allocated, then you may need to acquire a lock to allocate the storage. Some times it depends what libc you link against. What I read seemed to indicate that glibc's `pthread_getspecific` specifically is async-signal safe which is why I went this route.

Reply:
https://gcc.godbolt.org/z/3dTYsWxWj

There's no lazy allocation. The compiler puts thread-local definitions into a special ELF binary section that is initialized by the binary loader. `thread_local` is pretty much ideal on ELF platforms.

It might be different on Mac, but this code is Linux-specific anyway, so I wouldn't worry about that.

Reply:
I think that may not be true for code loaded from a shared library. So maybe we just document it as unsafe for dynamic linkage?

Reply:
Hmm you are right! If I add `-fPIC` to the godbolt example, it starts calling a function `__tls_get_addr`. Luckily, `-fPIE` does not do this. It's also apparently possible to use `__attribute((tls_model("initial-exec")))` on the thread-local variable, which avoids the call to `__tls_get_addr` with the caveat that the library cannot be dynamically `dlopen()`ed at runtime; it must be loaded strictly at program start.

Apparently, `__tls_get_addr` itself _could_ be signal-safe -- [apparently it is in musl libc](https://sourceware.org/legacy-ml/libc-alpha/2013-09/msg00569.html). But it seems glibc's version has not always been signal-safe, and while I can find evidence of people wanting to fix that, I can't find any definitive answer as to whether they did.

For our purposes, we don't care about `-fPIC`. But, you could throw `__attribute((tls_model("initial-exec")))` on the thread-locals anyway, so that if someone does try to turn on this feature with `-fPIC`, it'll probably still work. If they try to `dlopen()` the library, then the'll get an error. This feature isn't on by default anyway, so that doesn't seem like a big deal.

Reply:
Yup. That's what I had come across as well. `thread_local` + the attribute I think is the much better approach.